### PR TITLE
Chore: Refactor render via http

### PR DESCRIPTION
**Pull request by evictorero**
**What is this feature?**

Refactor render PNG, PDF and CSV
* Extract duplicated code
* Remove Encoding hardcoded value and use the renderType
* Rename functions
* Improve funcs responsibility

We could improve the codebase more, but this is big enough to review.

**Special notes for your reviewer:**
[Enterprise PR](https://github.com/grafana/grafana-enterprise/pull/6390)

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/rendering/http_mode.go [hunk 1]
```diff
@@ -171,20 +174,13 @@ func (rs *RenderingService) doRequest(ctx context.Context, u *url.URL, headers m
...
 
-func (rs *RenderingService) readFileResponse(ctx context.Context, resp *http.Response, filePath string, url string) error {
+func (rs *RenderingService) writeResponseToFile(ctx context.Context, resp *http.Response, filePath string) error {
 	// check for timeout first
 	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
 		rs.log.Info("Rendering timed out")
 		return ErrTimeout
 	}
 
-	// if we didn't get a 200 response, something went wrong.
```
Comment (evictorero) [inline pkg/services/rendering/http_mode.go:181]: Remove this block. This function shouldn't be called with an invalid resp. This error is not related to the func responsibility.

File: pkg/services/rendering/http_mode.go [hunk 2]
```diff
@@ -171,20 +174,13 @@ func (rs *RenderingService) doRequest(ctx context.Context, u *url.URL, headers m
 	return resp, nil
 }
 
-func (rs *RenderingService) readFileResponse(ctx context.Context, resp *http.Response, filePath string, url string) error {
+func (rs *RenderingService) writeResponseToFile(ctx context.Context, resp *http.Response, filePath string) error {
```
Comment (evictorero) [inline pkg/services/rendering/http_mode.go:177]: This name better represents the function intent.

File: pkg/services/rendering/interface.go [hunk 3]
```diff
@@ -42,18 +42,25 @@ func getRequestTimeout(opt TimeoutOpts) time.Duration {
 	return opt.Timeout * opt.RequestTimeoutMultiplier
 }
 
-type Opts struct {
+type CommonOpts struct {
```
Comment (evictorero) [inline pkg/services/rendering/interface.go:45]: I've extracted common opts. We could use `Opts` instead of having different structs, but I did this as a first step since I want to avoid changing all returning types.

File: pkg/services/rendering/plugin_mode.go [hunk 4]
```diff
@@ -9,10 +9,6 @@ import (
 )
 
 func (rs *RenderingService) renderViaPlugin(ctx context.Context, renderType RenderType, renderKey string, opts Opts) (*RenderResult, error) {
-	if renderType == RenderPDF {
-		opts.Encoding = "pdf"
-	}
```
Comment (evictorero) [inline pkg/services/rendering/plugin_mode.go:14]: I prefer not to override a parameter and keeping this sync is hard. I'm using the `renderType` as the encoding value below.

File: pkg/services/rendering/rendering.go [hunk 5]
```diff
@@ -264,7 +265,7 @@ func (rs *RenderingService) Render(ctx context.Context, renderType RenderType, o
 	result, err := rs.render(ctx, renderType, opts, renderKeyProvider)
 
 	elapsedTime := time.Since(startTime).Milliseconds()
-	saveMetrics(elapsedTime, err, RenderPNG)
+	saveMetrics(elapsedTime, err, renderType)
```
Comment (evictorero) [inline pkg/services/rendering/rendering.go:268]: This `Render` function is being called with other types not only `RenderPNG`

File: pkg/services/rendering/rendering.go [hunk 6]
```diff
@@ -296,7 +297,7 @@ func (rs *RenderingService) render(ctx context.Context, renderType RenderType, o
 		return rs.renderUnavailableImage(), nil
 	}
 
-	if renderType == RenderPDF || opts.Encoding == "pdf" {
```
Comment (evictorero) [inline pkg/services/rendering/rendering.go:299]: I couldn't find a case where we could receive a "pdf" encoding with a different `renderType` of PDF

File: pkg/services/rendering/interface.go [hunk 7]
```diff
@@ -75,14 +82,9 @@ type SanitizeSVGResponse struct {
...
 
-type CSVOpts struct {
-	TimeoutOpts
-	AuthOpts
-	Path            string
-	Encoding        string
-	Timezone        string
-	ConcurrentLimit int
-	Headers         map[string][]string
+type Result struct {
```
Comment (evictorero) [inline pkg/services/rendering/interface.go:85]: Generic result to reuse in all cases. I didn't change `RenderResult ` and `RenderCSVResult` to avoid changing all returning types.

File: pkg/services/rendering/http_mode.go [hunk 8]
```diff
@@ -36,109 +36,112 @@ var (
 	remoteVersionRefreshInterval               = time.Minute * 15
 )
 
+// renderViaHTTP renders PNG or PDF via HTTP
 func (rs *RenderingService) renderViaHTTP(ctx context.Context, renderType RenderType, renderKey string, opts Opts) (*RenderResult, error) {
```
Comment (evictorero) [inline pkg/services/rendering/http_mode.go:40]: This is the biggest change `renderViaHTTP` and `renderCSVViaHTTP` were almost identical. I've extracted all duplicated code. They could be the same function but it will be a bigger change because of returning and parameters types are different.

File: pkg/services/rendering/rendering.go [hunk 9]
```diff
@@ -406,7 +407,8 @@ func (rs *RenderingService) getNewFilePath(rt RenderType) (string, error) {
 	return filepath.Abs(filepath.Join(folder, fmt.Sprintf("%s.%s", rand, ext)))
 }
 
-func (rs *RenderingService) getURL(path string) string {
+// createURLForRendering creates a URL to send to the image rendering as callback for rendering a Grafana resource
+func (rs *RenderingService) createURLForRendering(path string) string {
```
Comment (AgnesToulet) [inline pkg/services/rendering/rendering.go:411]:
```suggestion
// getGrafanaURLForRendering creates a URL to send to the image rendering as callback for rendering a Grafana resource
func (rs *RenderingService) getGrafanaURLForRendering(path string) string {
```
I agree this can be improved but I was still confused with the new name so here are some suggestions to   clarify it. I think it would be nice to mention "Grafana" as this is the Grafana URL we are building and not the image renderer one (and that's the thing that can be confusing in the image renderer process).

Comment (AgnesToulet) [inline pkg/services/rendering/rendering.go:411]:
```suggestion
// getGrafanaCallbackURL creates a URL to send to the image rendering as callback for rendering a Grafana resource
func (rs *RenderingService) getGrafanaCallbackURL(path string) string {
```

**Review by AgnesToulet — COMMENTED** [review]
Great refactor! I left a few minor suggestions. My main concern is the render API where `encoding` can be set to `pdf` to render a PDF and this should impact the `renderType`.

File: pkg/api/render.go [hunk 10]
```diff
@@ -59,22 +59,23 @@ func (hs *HTTPServer) RenderToPng(c *contextmodel.ReqContext) {
...
+			Path:            web.Params(c.Req)["*"] + queryParams,
+			Timezone:        queryReader.Get("tz", ""),
+			ConcurrentLimit: hs.Cfg.RendererConcurrentRequestLimit,
+			Headers:         headers,
 		},
 		Width:             width,
 		Height:            height,
-		Path:              web.Params(c.Req)["*"] + queryParams,
-		Timezone:          queryReader.Get("tz", ""),
-		Encoding:          encoding,
```
[the remark this answers is not in the document]
Comment (evictorero) [reply pkg/api/render.go:74]: This confuses me a bit. The function name is `RenderToPng`, and the `RenderService.Render` function is being called with a hardcoded type `rendering.RenderPNG`. That is why I removed it. If this could be called using 'pdf,' then we should think of changing names and confirm that changing the render type above on line 61 has no side effect.

[the remark this answers is not in the document]
Comment (evictorero) [reply pkg/api/render.go:74]: I've made the changes. Now the `renderType` is inferred by the encoding param and managed by the API consumer like before. WDYT? I don't want to change the behavior in this PR.

**Review by AgnesToulet — APPROVED** [review]
